### PR TITLE
gh-85984: Utilize new "winsize" functions from termios in pty tests.

### PR DESCRIPTION
**Pull request by 8vasu**
This follows https://github.com/python/cpython/pull/92365. This is one in a series of PRs aimed at cleaning-up, fixing bugs in, introducing new features in, and updating the code in "Lib/pty.py".

This is a dependency of https://github.com/python/cpython/pull/101833. 

https://github.com/python/cpython/pull/101832 is another dependency of https://github.com/python/cpython/pull/101833.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

* Issue: gh-85984

### CODE REVIEW

**Comment by 8vasu** [comment]
@gpshead Thank you for being a major part of this project for so long. This is another PR for you. There is no rush; please take a look at this when you are free.

File: Lib/test/test_pty.py [hunk 1]
```diff
@@ -409,27 +387,27 @@ def test__copy_to_each(self):
...
-        pty.select = self._mock_select
-        self.select_rfds_lengths.append(2)
-        self.select_rfds_results.append([mock_stdin_fd, masters[0]])
-        # We expect that both fds were removed from the fds list as they
-        # both encountered an EOF before the second select call.
-        self.select_rfds_lengths.append(0)
-
-        # We expect the function to return without error.
-        self.assertEqual(pty._copy(masters[0]), None)
+    # def test__copy_eof_on_all(self):
```
Comment (8vasu) [inline Lib/test/test_pty.py]: `test__copy_eof_on_all` probably does not make sense due to the current structure of the `_copy()` function. Please correct me if I am wrong.

File: Lib/test/test_pty.py [hunk 2]
```diff
@@ -15,20 +15,12 @@
 import socket
 import io # readline
 import unittest
-
-import struct
-import fcntl
```
Comment (gpshead) [inline Lib/test/test_pty.py:20]: there's a test.support.import_helper.import_module("fcntl") on line 6 as well to skip the test if it doesn't exist, does this test have reason to care about fcntl existing at all now?

File: Lib/test/test_pty.py [hunk 3]
```diff
@@ -105,18 +89,14 @@ def setUp(self):
...
             try:
-                stdin_dim = os.get_terminal_size(pty.STDIN_FILENO)
-                self.stdin_rows = stdin_dim.lines
-                self.stdin_cols = stdin_dim.columns
-                old_stdin_winsz = struct.pack("HHHH", self.stdin_rows,
-                                              self.stdin_cols, 0, 0)
-                self.addCleanup(_set_term_winsz, pty.STDIN_FILENO, old_stdin_winsz)
-            except OSError:
+                self.stdin_dim = tty.tcgetwinsize(pty.STDIN_FILENO)
+                self.addCleanup(tty.tcsetwinsize, pty.STDIN_FILENO, \
```
Comment (gpshead) [inline Lib/test/test_pty.py]: unnecessary backslash

File: Lib/test/test_pty.py [hunk 4]
```diff
@@ -131,41 +111,40 @@ def test_openpty(self):
...
         except tty.error:
-            # not a tty or bad/closed fd
+            # Not a tty or bad/closed fd.
             debug("tty.tcgetattr(pty.STDIN_FILENO) failed")
             mode = None
 
-        new_stdin_winsz = None
-        if self.stdin_rows is not None and self.stdin_cols is not None:
+        new_dim = None
+        if self.stdin_dim != None:
```
Comment (gpshead) [inline Lib/test/test_pty.py]: just write this as `if self.stdin_dim:` there i no need to check specifically for None.

File: Lib/test/test_pty.py [hunk 5]
```diff
@@ -131,41 +111,40 @@ def test_openpty(self):
...
             except OSError:
-                warnings.warn("Failed to set pty.STDIN_FILENO window size")
+                warnings.warn("Failed to set pty.STDIN_FILENO window size.")
                 pass
 
         try:
             debug("Calling pty.openpty()")
             try:
-                master_fd, slave_fd = pty.openpty(mode, new_stdin_winsz)
+                master_fd, slave_fd, slave_name = pty.openpty(mode, new_dim, \
```
Comment (gpshead) [inline Lib/test/test_pty.py]: no need for the backslash \  - in general, always avoid backslash line continuation as a style rule.  within any form of parens or brackets it is unnecessary.

File: Lib/test/test_pty.py [hunk 6]
```diff
@@ -131,41 +111,40 @@ def test_openpty(self):
...
+                                                              True)
             except TypeError:
                 master_fd, slave_fd = pty.openpty()
-            debug(f"Got master_fd '{master_fd}', slave_fd '{slave_fd}'")
+                slave_name = None
+            debug(f"Got master_fd '{master_fd}', slave_fd '{slave_fd}', slave_name '{slave_name}'")
```
Comment (gpshead) [inline Lib/test/test_pty.py]: modern f-strings have a trailing = for debug style output, so you don't need to repeat information, use something like:

debug(f"Got {master_fd=}, {slave_fd=}, {slave_name=}")

that'll render with the variable names and their values automatically.

File: Lib/test/test_pty.py (hunk 1 above)
Comment (gpshead) [reply Lib/test/test_pty.py]: agreed, this test doesn't make sense as written, please delete it.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Lib/test/test_pty.py (hunk 2 above)
Comment (8vasu) [reply Lib/test/test_pty.py:20]: Good catch.

Some context: I was the one who added the original `import struct` and `import fcntl` to define makeshift `getwinsize` and `setwinsize` type functions. But now that we have `termios.tcgetwinsize()` and `termios.tcsetwinsize()`, we no longer need those.

Answer to your original question: I think @tiran added the `import_module("fcntl")` on line 6 in the following commit to skip these tests on `wasm32-emscripten`: https://github.com/python/cpython/commit/c02e860ee79f29905be6fca997c96bb1a404bb32

I will remove that `import_module("fcntl")` line now.

File: Lib/test/test_pty.py (hunk 5 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Thank you for teaching me this. Removing.

File: Lib/test/test_pty.py (hunk 3 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Removing.

File: Lib/test/test_pty.py (hunk 4 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Makes sense.

File: Lib/test/test_pty.py (hunk 6 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Awesome! Did not know this either.

File: Lib/test/test_pty.py (hunk 1 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Done.

**Comment by 8vasu** [comment]
I have made the requested changes; please review again

**Comment by gpshead** [comment]
!buildbot wasm32

File: Lib/test/test_pty.py (hunk 2 above)
Comment (gpshead) [reply Lib/test/test_pty.py:20]: sadly the wasm32 tests crash on this test, so i've restored the line with a comment explaining why for now.

**Comment by gpshead** [comment]
!buildbot wasm32

**Comment by gpshead** [comment]
_(specifically it was the wasm32 encripten node bots that crashed on test_pty)_

**Comment by 8vasu** [comment]
@gpshead Thank you for the final change. Some info: maybe the reason why it failed with both the Python `fcntl` module and the current `termios.tc{get/set}winsize()` functions is that `ioctl()+TIOCGWINSZ/TIOCSWINSZ` is problematic on wasm32-emcripten. When I googled "emscripten ioctl", this was the very second link: https://github.com/emscripten-core/emscripten/issues/6463.

**Pull request merged**